File: src/geegoo_agent/llm/presets.py
"""Built-in LLM provider presets (OpenAI-compatible APIs)."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

from geegoo_agent.llm.openai_provider import OpenAIProvider

LLMProviderName = Literal["openai", "deepseek", "minimax"]


@dataclass(frozen=True)
class ProviderPreset:
    name: LLMProviderName
    label: str
    base_url: str | None
    default_model: str


PROVIDER_PRESETS: dict[LLMProviderName, ProviderPreset] = {
    "openai": ProviderPreset(
        name="openai",
        label="OpenAI",
        base_url=None,
        default_model="gpt-4o",
    ),
    "deepseek": ProviderPreset(
        name="deepseek",
        label="DeepSeek",
        base_url="https://api.deepseek.com",
        default_model="deepseek-v4-flash",
    ),
    "minimax": ProviderPreset(
        name="minimax",
        label="Minimax",
        base_url="https://api.minimaxi.com/v1",
        default_model="MiniMax-M2.1",
    ),
}

# (model_id, short description) — see https://api-docs.deepseek.com/zh-cn/
PROVIDER_MODELS: dict[LLMProviderName, list[tuple[str, str]]] = {
    "deepseek": [
        ("deepseek-v4-flash", "V4 Flash，快速对话（推荐默认）"),
        ("deepseek-v4-pro", "V4 Pro，复杂推理 / 思考模式"),
        ("deepseek-chat", "旧版 chat（2026/07 弃用，兼容）"),
        ("deepseek-reasoner", "旧版 reasoner（2026/07 弃用，兼容）"),
    ],
    "openai": [
        ("gpt-4o", "GPT-4o"),
        ("gpt-4o-mini", "GPT-4o mini"),
        ("gpt-4-turbo", "GPT-4 Turbo"),
    ],
    "minimax": [
        ("MiniMax-M2.1", "MiniMax M2.1"),
    ],
}


def list_provider_models(provider: LLMProviderName) -> list[tuple[str, str]]:
    return list(PROVIDER_MODELS.get(provider, []))


def current_model(provider: LLMProviderName, model: str | None = None) -> str:
    return resolve_model(provider, model)


def pick_model(provider: LLMProviderName, choice: str, *, current: str | None = None) -> str:
    """Resolve ``/model`` selection: index (1-based), model id, or default."""
    text = choice.strip()
    if not text:
        return current_model(provider, current)

    models = list_provider_models(provider)
    if text.isdigit():
        index = int(text)
        if 1 <= index <= len(models):
            return models[index - 1][0]
        raise ValueError(f"invalid model index: {index} (1-{len(models)})")

    known = {model_id for model_id, _ in models}
    if known and text not in known:
        raise ValueError(f"unknown model for {provider}: {text}")
    return text


def resolve_model(provider: LLMProviderName, model: str | None = None) -> str:
    preset = PROVIDER_PRESETS[provider]
    return (model or "").strip() or preset.default_model


def model_supports_thinking(provider: LLMProviderName, model: str | None = None) -> bool:
    if provider != "deepseek":
        return False
    name = resolve_model(provider, model).lower()
    return "v4" in name or name == "deepseek-reasoner"


def resolve_thinking_enabled(
    provider: LLMProviderName,
    model: str | None,
    *,
    thinking: bool | None,
) -> bool:
    if not model_supports_thinking(provider, model):
        return False
    if thinking is None:
        name = resolve_model(provider, model).lower()
        return "v4" in name
    return thinking


def build_llm_provider(
    provider: LLMProviderName,
    token_key: str,
    *,
    model: str | None = None,
    thinking: bool | None = None,
    reasoning_effort: str = "high",
) -> OpenAIProvider:
    preset = PROVIDER_PRESETS[provider]
    resolved_model = resolve_model(provider, model)
    thinking_enabled = resolve_thinking_enabled(
        provider, resolved_model, thinking=thinking
    )
    return OpenAIProvider(
        resolved_model,
        token_key,
        base_url=preset.base_url,
        thinking_enabled=thinking_enabled,
        reasoning_effort=reasoning_effort if thinking_enabled else None,
    )

File: internal/llm/presets.go
package llm

import (
	"fmt"
	"strconv"
	"strings"
)

// ProviderName identifies built-in LLM backends.
type ProviderName string

const (
	ProviderOpenAI   ProviderName = "openai"
	ProviderDeepSeek ProviderName = "deepseek"
	ProviderMinimax  ProviderName = "minimax"
)

// Preset holds default base URL and model for a provider.
type Preset struct {
	Name         ProviderName
	Label        string
	BaseURL      string
	DefaultModel string
}

// Presets maps provider name to preset.
var Presets = map[ProviderName]Preset{
	ProviderOpenAI: {
		Name:         ProviderOpenAI,
		Label:        "OpenAI",
		BaseURL:      "https://api.openai.com/v1",
		DefaultModel: "gpt-4o",
	},
	ProviderDeepSeek: {
		Name:         ProviderDeepSeek,
		Label:        "DeepSeek",
		BaseURL:      "https://api.deepseek.com",
		DefaultModel: "deepseek-v4-flash",
	},
	ProviderMinimax: {
		Name:         ProviderMinimax,
		Label:        "Minimax",
		BaseURL:      "https://api.minimaxi.com/v1",
		DefaultModel: "MiniMax-M2.1",
	},
}

// ProviderModel is one selectable model for /model.
type ProviderModel struct {
	ID          string
	Description string
}

// ProviderModels lists known models per provider (Python PROVIDER_MODELS).
var ProviderModels = map[ProviderName][]ProviderModel{
	ProviderDeepSeek: {
		{ID: "deepseek-v4-flash", Description: "V4 Flash，快速对话（推荐默认）"},
		{ID: "deepseek-v4-pro", Description: "V4 Pro，复杂推理 / 思考模式"},
		{ID: "deepseek-chat", Description: "旧版 chat（2026/07 弃用，兼容）"},
		{ID: "deepseek-reasoner", Description: "旧版 reasoner（2026/07 弃用，兼容）"},
	},
	ProviderOpenAI: {
		{ID: "gpt-4o", Description: "GPT-4o"},
		{ID: "gpt-4o-mini", Description: "GPT-4o mini"},
		{ID: "gpt-4-turbo", Description: "GPT-4 Turbo"},
	},
	ProviderMinimax: {
		{ID: "MiniMax-M2.1", Description: "MiniMax M2.1"},
	},
}

// ListProviderModels returns models for a provider.
func ListProviderModels(name ProviderName) []ProviderModel {
	return ProviderModels[name]
}

// ResolveModel returns configured model or provider default.
func ResolveModel(name ProviderName, model string) string {
	if strings.TrimSpace(model) != "" {
		return strings.TrimSpace(model)
	}
	if p, ok := Presets[name]; ok {
		return p.DefaultModel
	}
	return "gpt-4o"
}

// PickModel resolves /model selection: 1-based index, model id, or default.
func PickModel(name ProviderName, choice, current string) (string, error) {
	text := strings.TrimSpace(choice)
	if text == "" {
		return ResolveModel(name, current), nil
	}
	models := ListProviderModels(name)
	if n, err := strconv.Atoi(text); err == nil {
		if n >= 1 && n <= len(models) {
			return models[n-1].ID, nil
		}
		return "", fmt.Errorf("invalid model index: %d (1-%d)", n, len(models))
	}
	if len(models) > 0 {
		known := map[string]struct{}{}
		for _, m := range models {
			known[m.ID] = struct{}{}
		}
		if _, ok := known[text]; !ok {
			return "", fmt.Errorf("unknown model for %s: %s", name, text)
		}
	}
	return text, nil
}

// ModelSupportsThinking reports whether thinking mode applies.
func ModelSupportsThinking(name ProviderName, model string) bool {
	if name != ProviderDeepSeek {
		return false
	}
	resolved := strings.ToLower(ResolveModel(name, model))
	return strings.Contains(resolved, "v4") || resolved == "deepseek-reasoner"
}

// ResolveThinkingEnabled mirrors Python resolve_thinking_enabled.
func ResolveThinkingEnabled(name ProviderName, model string, thinking *bool) bool {
	if !ModelSupportsThinking(name, model) {
		return false
	}
	if thinking == nil {
		return strings.Contains(strings.ToLower(ResolveModel(name, model)), "v4")
	}
	return *thinking
}
